### ISSUE TITLE
Add J

### ISSUE THREAD

**Issue by RustemB** [codewars/codewars-runner-cli#821]
- Name: J
- Website: https://www.jsoftware.com/
- Language Version: `j807`, `j901`

- Test Frameworks: just load file/script and use needed functions
- How to install:
```bash
# for j901
$ wget http://www.jsoftware.com/download/j901/install/j901_amd64.deb
$ sudo dpkg -i j901_amd64.deb
# for j807
$ wget http://www.jsoftware.com/download/j807/install/j807_amd64.deb
$ sudo dpkg -i j807_amd64.deb
```

**Comment by Steffan153** [comment]
This would be interesting, but I think it would be tricky to get a working setup.

**Comment by g964** [comment]
The author will have to write tests using some J primitives since there are no testing framework. CW could concatenate the solution and the tests and run the whole but I know nothing about the CW runner. I wrote something from the "multiply" kata for J :
```
multiply =: 4 : 0
   x * y
)

testing =: 3 : 0
   t =. (actual = expect)
   if. t = 0 do. 
        echo 'Incorrect value, expected is: '
        echo expect
        echo 'but got: '
        echo actual
   else.
      echo 'correct'
      echo 'actual'
      echo actual
      echo 'expect'
      echo expect
   end.
   assert. t = 1
)
fixedtests =: 3 : 0 
   actual =: 5 multiply 8
   expect =: 40
   testing ''
   echo '---------------'
   actual =: 8 multiply 9
   expect =: 72
   testing ''
   echo '---------------'
   actual =: 10 multiply 1000
   expect =: 9999
   testing ''
   echo 'End of fixed tests'
)

fixedtests ''
```
The last test fails and tests are stopped without echoing 'End of fixed tests'.
The solution part is the function "multiply" and after that the tests part. You can try it at "tio" with the first two parts in "code" and the "fixedtests ''" part in input. You get the result in "output".

**Comment by RustemB** [comment]
There are many other ways to produce test units

**Comment by g964** [comment]
@RustemB: could be good if you gave something with details:-)

**Comment by Steffan153** [comment]
We could write our own test framework and just include it, right?

**Comment by kazk** [comment]
Yes, anything should work as long as it produces the Codewars output. We should use the existing tools when possible though.
How does J developers write tests?

**Comment by Bubbler-4** [comment]
> How does J developers write tests?

Usually it's done using the built-in assertion primitive `assert.`, but on assertion failure it gives a fixed output:

```j
assert. 1 = 0
|assertion failure: examplePgm
|   1    =0
```

(And it can be turned off using system functions.)

So I think we will need a custom test framework.

**Comment by kazk** [comment]
I found this: https://github.com/jsoftware/general_unittest

Can we fork that and customize the test output?
https://github.com/jsoftware/general_unittest/blob/c0916768bba3832fbb9bf2305d34575a5adc0ad0/unittest.ijs#L21-L62

But I guess we can only produce top level assertion results without test names with that.

---

> So I think we will need a custom test framework. But we have a special challenge here because J does not allow multi-line statements or nesting multi-line definitions. Do we have other languages with similar constraints?

Can you elaborate? Do you mean it's not easy to have the test name and the test duration because of J's limitation?

**Comment by Bubbler-4** [comment]
I thought it's not easy to visually simulate how the test cases are written in other languages. That unittest library does look promising (though it might still need a complete rewrite for our needs, it's indeed worth taking a look).
